[RFC] Expanding Moteus Communication for Debug

lucienpeach opened the pull request:
Introduces new accessor functions within `joint_moteus.h` to allow for inclusion of additional registers [as defined](https://github.com/mjbots/moteus/blob/main/docs/reference.md#a2b-registers) in the moteus controller `reference.md`. Includes supporting architecture via `options` to allow user specification of desired registers to be read in controller definition. 

### **Change List**

- Provides new accessor functions to request additional registers from the moteus, as defined in the `QueryCommand` struct within `moteus_protocol.h`. These additional registers include:
  - `q_current`, the current in the Q phase measured in amperes.
  - `d_current`, the current in the D phase measured in amperes.
  - `temperature`, the current board temperature, measured in degrees Celsius.
  - `voltage`, the current input voltage.
  - `fault`, a fault code which will be set if the primary mode is 1 (Fault).
- Adds fault code enumeration (pulled from [fw/error.h](https://github.com/mjbots/moteus/blob/main/fw/error.h#L25) from the moteus repository) to enable reading of fault code information, if desired.
- Adds joint vector getter `GetJoints()` to `mjbots_hardware_interface.h` which returns a vector of shared pointers to objects of the `JointMoteus` class. This allows the new accessor functions in `joint_moteus.h` to be referenced from the controller level to allow for logging of the respective registers. These new accessor functions include:
  - `get_temperature()`, which provides the `temperature` register.
  - `get_q_current()`, which provides the `q_current` register.
  - `get_d_current()`, which provides the `d_current` register.
  - `get_voltage()`, which provides the `voltage` register.
  - `get_fault()`, which provides the `fault` register.
- Provides new boolean options within the `MjbotsHardwareInterface` class which dictate the request sent to the moteus by the pi3hat and the register packets to be sent from the moteus to the pi3hat. These new options include:
  -  `send_current`, which includes packets from `q_current` and `d_current` registers if true. 
  - `send_voltage`, which includes packets from `voltage` register if true.
  - `send_temperature`, which includes packets from `temperature` register if true.
  - `send_fault`, which includes packets from `fault` register if true. 

### **Upcoming Changes**

- Create a `joint_moteus.cpp` implementation file to clean up `joint_moteus.h` and prevent this file from becoming too convoluted.
- Streamline code to eliminate static resolution definition and reduce the number of options calls required. This would allow the user to specify the resolution of a register while simultaneously identifying that register to be included in packets from the moteus. 
  - This proposed change would require numerous type-related changes and edits within a number of low-level files, the specifics of which ought to be a topic of discussion and deliberation. 

### **Testing**

Hardware testing on the DOQ robot is planned to verify the effectiveness of this code in practice. This testing will primarily seek to validate logging of new registers from the moteus.

### **Questions**

There are a number of ways to build off this code and optimize it. Discussion of changes proposed above and suggestions on additional new changes are greatly appreciated.

lucienpeach committed:
Moteus Registers Debug Options

Initial commit for moteus-register-debug-data branch. Introduce additional options to send moteus packages for logging and debugging.

lucienpeach committed:
Moteus Registers Debug Additions

Additional ternary options for UpdateMoteus(). Debug register additions in joint_moteus.h. Fault codes added to moteus_protocol.h.

lucienpeach committed:
Remove erroneous comment

Copy and paste error comment from the other joint getters

lucienpeach committed:
Clarify comment, value is board temperature

lucienpeach committed:
Add bool arguments for debug options

Constructor was missing necessary boolean options, causing compilation issues.

lucienpeach committed:
Add joint vector getter to hardware interface

lucienpeach committed:
Added Boolean Options to mjbots_control_loop.h

jdcaporale commented on include/kodlab_mjbots_sdk/joint_moteus.h:
```diff
@@ -129,20 +157,43 @@ class JointMoteus: public JointBase
         
         /**
-         * @brief Update the joint of the moteus. Converts rot/s to rad/s and saves mode
+         * @brief Update the joint of the moteus. Converts rot/s to rad/s and saves registers
```
I would change this to state values 
```suggestion
         * @brief Update the joint of the moteus. Converts rot/s to rad/s and saves 
         other state values
```

jdcaporale commented on include/kodlab_mjbots_sdk/joint_moteus.h:
```diff
@@ -129,20 +157,43 @@ class JointMoteus: public JointBase
+         * @param reply_d_current current in the D phase reported by moteus [A]
+         * @param reply_voltage voltage reported by the moteus [V]
+         * @param reply_temperature temperature reported by the moteus [C]
+         * @param fault     fault code
```
```suggestion
         * @param fault fault code
```

jdcaporale commented on include/kodlab_mjbots_sdk/joint_moteus.h:
```diff
@@ -129,20 +157,43 @@ class JointMoteus: public JointBase
-            mode_ = mode;
-        }
+        void UpdateMoteus(
+            float reply_pos,
+            float reply_vel,
+            float reply_torque,
+            ::mjbots::moteus::Mode mode,
+            float reply_q_current,
+            float reply_d_current,
+            float reply_voltage,
+            float reply_temperature,
+            ::mjbots::moteus::Fault fault){
```
Should we just take in the reply itself and parse some here?

ShaneRozenLevy replied:
That really depends on how we are scoping, and I think both options are reasonable.

jdcaporale commented on src/mjbots_hardware_interface.cpp:
```diff
@@ -43,6 +43,19 @@ void MjbotsHardwareInterface::InitializeCommand() {
+    if(send_temperature_){
+      cmd.query.temperature = ::mjbots::moteus::Resolution::kInt16;
+    }
+    if(send_fault_){
+      cmd.query.fault = ::mjbots::moteus::Resolution::kInt8;
+    }
```
Should we always ask for this, or only when the mode returns a fault?

jdcaporale replied:
e.g pseudo code

> 
> if( mode == FAULT and send_fault_) {
>   cmd.query.fault = kInt8
> } else{
>   cmd.query.fault = kIgnore}

jdcaporale replied:
depends on how the mode message works

jdcaporale replied:
After talking about it, we shouldn't change this

jdcaporale commented on src/mjbots_hardware_interface.cpp:
```diff
@@ -70,7 +83,11 @@ MjbotsHardwareInterface::MjbotsHardwareInterface(std::vector<std::shared_ptr<Joi
                                                  bool dry_run,
                                                  bool print_torques,
-                                                 bool send_pd_commands)
+                                                 bool send_pd_commands,
+                                                 bool send_current,
+                                                 bool send_voltage,
+                                                 bool send_temperature,
+                                                 bool send_fault)
```
maybe change bool names to request_XXXX except for send_pd_commands, which is actually a set of values.

jdcaporale reviewed (commented):
High level comments: looks good!

- Update Moteus should take in a reply object
- maybe a good place for the use of std::optional? we should discuss
- A MessageRequestOptions sub structt might be nice to pass around
- I don't know if we need to constantly read fault, depending on how the mode message works

ShaneRozenLevy commented on include/kodlab_mjbots_sdk/joint_moteus.h:
```diff
@@ -33,6 +33,10 @@ struct MoteusJointConfig{
     float moteus_kd = 0;
     float soft_start_duration_ms = 1;
+    float temperature = 0;
+    float q_current = 0;
```
Why is a state in the config struct?

ShaneRozenLevy commented on include/kodlab_mjbots_sdk/joint_moteus.h:
```diff
@@ -71,12 +79,20 @@ class JointMoteus: public JointBase
-            float moteus_kd = 0)
+            float moteus_kd = 0,
+            float temperature = 0,
+            float q_current = 0,
```
Why is a state being set by the constructor?

jdcaporale replied:
I misunderstood what we were talking about when we were talking earlier and now understand why you were so flabbergasted and agree.

ShaneRozenLevy commented on include/kodlab_mjbots_sdk/joint_moteus.h:
```diff
@@ -165,6 +216,41 @@ class JointMoteus: public JointBase
+         * 
+         * @return float
+         */
+        virtual float get_temperature() const {return temperature_;}
```
Why are these virtual?

ShaneRozenLevy commented on include/kodlab_mjbots_sdk/mjbots_control_loop.h:
```diff
@@ -41,6 +41,10 @@ struct ControlLoopOptions {
   bool dry_run = false;  ///< If true, torques sent to moteus boards will always be zero
   bool print_torques = false;  ///< If true, torque commands will be printed to console
   bool send_pd_commands = false; ///< If true, the control loop will send pd setpoints & gains in addition to ffwd torque commands
+  bool send_current = false;  ///< If true, the control loop will send d_current and q_current registers
```
You should just have the resolution struct here instead of a bunch of bools.

ShaneRozenLevy reviewed (changes requested):
2 big issues are constructor for joint moteus, and the request data should be set via an option that is `QueryCommand` instead of the bools.

lucienpeach committed:
joint_moteus.cpp Source File Addition

Edited joint_moteus.h and created new corresponding source file joint_moteus.cpp. Added preliminary comments for user QueryCommand struct entry.

lucienpeach committed:
QueryCommand user library additions

Added library of preset QueryCommand structs for common register requests.

lucienpeach committed:
UpdateMoteus() and boolean options streamlining

Implemented supporting architecture for QueryCommand struct input. Modified UpdateMoteus() to accept entire reply_message.

lucienpeach committed:
Change  to

lucienpeach committed:
Changed torque resolutions to kInt32 within QueryCommand struct library

lucienpeach committed:
Changed feedforward_torque resolution in InitializeCommand() to kInt32

lucienpeach committed:
Undo change on commanded feedforward_torque res, adjust commanded max torque res for use_pd_commands

lucienpeach committed:
LOG_WARN implementation and override functions at joint_moteus level

lucienpeach committed:
Open loop control permissions and control safety features

Adds a boolean open_loop flag for user specification of open loop control and safety checks for improper closed loop operation

lucienpeach committed:
Bolster open loop functionality

Prevents unnecessary errors during open loop, adds safety features for closed loop, adds get_mode() to joint_moteus

lucienpeach commented:
Builds upon previous changes to allow requests to be sent via a `QueryCommand` struct that replaces the previous boolean method of requesting data while bolstering user optionality through resolution selection. Includes initial support for open-loop control and introduces safety checks via warning messages and `kodlab::ActivateCtrlC()` to alert the user of input resolution errors and to prevent unsafe and improper closed-loop control.

### **Change List**

- Introduces a source file `joint_moteus.cpp` to streamline header vs. implementation for the `JointMoteus` class. 
- Adds a library of preset `QueryCommand` structs at the `JointMoteus` class level for common combinations of requests and resolutions. This library is not comprehensive and thus the user can make adjustments at the controller level as desired. 
- Condensed `UpdateMoteus()` to accept a `reply_message` object of type `QueryResult`, as opposed to requiring a large number of individual reply fields (`reply_pos`, `reply_vel`, etc.) as inputs. This goes hand-in-hand with the previously stated change to allow for `QueryCommand` structs as user inputs.
- Because of the additional freedom provided to the user through the passing of a `QueryCommand` struct, it is now possible for a new and inexperienced user to encounter previously unseen issues. To prevent this from happening, and to inform the inexperienced user of these issues, a number of additions have been made:
  - Adds `LOG_WARN` messages within accessor functions at the `JointMoteus` class level which inform the user if the resolution of a requested data field is set to `kIgnore`. If the resolution is set to a different value, the accessor function provides the respective field at the resolution specified in the `QueryCommand` struct. These messages are of "warning" status and thus do not kill the control loop. 
  - Adds initial support for open-loop control via the introduction of a new option `open_loop` which is stored within the `options` struct. If this value is set to `true`, the control loop will not check to ensure that `position` and `velocity` resolutions are set to a value other than `kIgnore` (since these readings are not necessary for open-loop control). 
  - Introduces a new function in the `JointMoteus` class, `is_open_loop()`, which checks to ensure that the resolutions of `position` and `velocity` are set to values other than `kIgnore`. If either of these values are set to `kIgnore` in the `QueryCommand` input struct, `is_open_loop()` returns a boolean value of `true`, indicating that the current resolution configuration requires open-loop control. 
  - Outputs a `LOG_FATAL` message and runs `kodlab::ActivateCtrlC()` if `options.open_loop` is set to `false` but `is_open_loop()` outputs a value of `true`, indicating that open-loop control is necessary for the current resolution configuration but not yet specified. This effectively kills the control loop from within the constructor definition, preventing the user from running closed-loop control without reporting position or velocity. 
- Provides new accessor functions in the `JointMoteus` class:
  - `get_velocity()`, which overrides the respective virtual function in the `JointBase` class and enables `LOG_WARN` message support.
  - `get_position()`, which overrides the respective virtual function in the `JointBase` class and enables `LOG_WARN` message support.
  - `get_servo_torque()`, which overrides the respective virtual function in the `JointBase` class and enables `LOG_WARN` message support.
  - `get_measured_torque()`, which overrides the respective virtual function in the `JointBase` class and enables `LOG_WARN` message support.
  - `get_mode()`, which allows allows the user to directly access the mode (circumventing the current architecture, which requires accessing `GetJointModes()` via the `MjbotsHardwareInterface` class). Additionally provides support for `LOG_WARN` messages.
  - `get_query_command()`, which allows the user to access the current `QueryCommand` struct.
- Introduces the `all_unset()` function within `QueryResult` in the `moteus` namespace, which is used to determine if the fields within a given `QueryResult` struct are populated or are set to default values. Added a new enum value `kUnset` to both `Mode` and `Fault` enums within the `moteus` namespace to represent these unset/default values (set to -1).
  - In conjunction with the `any_set()` function which exists within `QueryCommand` in the `moteus` namespace, these additions are intended to provide more accurate insight into potential drops in CAN communication via terminal messages.
- Minor changes to settings and naming conventions. 
  - Changed `send_pd_commands` bool to `use_pd_commands` to better encompass the utility of this option.
  - Adjusted maximum commanded torque resolution `res.maximum_torque` to `kInt16` resolution from `kInt8` to prevent cutoffs of 0.5 Nm when `use_pd_commands` is set to `true`.

### **Testing**
Testing on DOQ has confirmed proper communication with the robot and functionality of safety features in both open- and closed-loop control schemes (toggled via the `open_loop` option). Plotting has demonstrated reasonable values for all requested fields and control loop communications function without packet loss at 1000 Hz even with a fully comprehensive `QueryCommand` struct that returns all possible data fields. 

### **Questions** 
Open to suggestions on `QueryCommand` library changes or additions. Interested in thoughts of more robust open-loop support in future PRs.

lucienpeach committed:
Merge branch 'master' into moteus-register-debug-data

ShaneRozenLevy commented on include/kodlab_mjbots_sdk/joint_moteus.h:
```diff
@@ -1,7 +1,7 @@
  * @file joint_moteus.h
  * @author J. Diego Caporale (jdcap@seas.upenn.edu)
- * @brief Moteus powered joint class
+ * @brief Moteus powered joint class header
```
You don't need to change this. The header is for the class.

ShaneRozenLevy commented on include/kodlab_mjbots_sdk/joint_moteus.h:
```diff
@@ -43,6 +31,30 @@ struct MoteusJointConfig{
     public:
+        /**
+         * @brief Define a default resolution struct
+         * 
```
Fix the extra line here.

ShaneRozenLevy commented on include/kodlab_mjbots_sdk/joint_moteus.h:
```diff
@@ -43,6 +31,30 @@ struct MoteusJointConfig{
+         * @brief Define a default resolution struct
+         * 
+         */
+        static const ::mjbots::moteus::QueryCommand kDefaultQuery;
```
@jdcaporale Do you think this is the correct place for these to live? Sort of torn if they should live here or in a namespace.

lucienpeach replied:
Relocated to the bottom of the file for the time being, in keeping with the comparable change to source file.

jdcaporale replied:
I think this is fine.

ShaneRozenLevy commented on include/kodlab_mjbots_sdk/joint_moteus.h:
```diff
@@ -43,6 +31,30 @@ struct MoteusJointConfig{
+        static const ::mjbots::moteus::QueryCommand kTorqueQuery;
+
+        /**
+         * @brief Define a debug resolution struct
```
What is the difference between `kDebugQuery` and `kComprehensiveQuery`? Its not obvious from the comments.

ShaneRozenLevy commented on include/kodlab_mjbots_sdk/joint_moteus.h:
```diff
@@ -103,129 +114,193 @@ class JointMoteus: public JointBase
 
         /**
-         * @brief Update the joint of the moteus. Converts rot/s to rad/s and saves mode
+         * @brief Determines whether open loop options should be enabled
```
Is this function calculating if open loop, or just returning a bool? If its calculating, what determines if the moteus is open loop? Make that clear in the comment

lucienpeach replied:
Changed comment and function name to better describe the utility of this function. Open to additional changes/suggestions.

ShaneRozenLevy commented on include/kodlab_mjbots_sdk/joint_moteus.h:
```diff
@@ -103,129 +114,193 @@ class JointMoteus: public JointBase
+         * 
+         * @return const ::mjbots::moteus::QueryCommand
+         */
+        const ::mjbots::moteus::QueryCommand get_query_command() const;
```
Either return a const reference or return by value. Don't return by const value.

ShaneRozenLevy commented on include/kodlab_mjbots_sdk/joint_moteus.h:
```diff
@@ -103,129 +114,193 @@ class JointMoteus: public JointBase
+        /**
+         * @brief Get the mode 
+         * 
+         * @return const ::mjbots::moteus::Mode
```
These ai generated comments are not useful. Either make them useful or remove them.

ShaneRozenLevy commented on include/kodlab_mjbots_sdk/joint_moteus.h:
```diff
@@ -103,129 +114,193 @@ class JointMoteus: public JointBase
+         * 
+         * @return const ::mjbots::moteus::Mode
+         */
+        const ::mjbots::moteus::Mode get_mode() const;
```
These should probably be [[nodiscard]]

ShaneRozenLevy commented on include/kodlab_mjbots_sdk/joint_moteus.h:
```diff
@@ -103,129 +114,193 @@ class JointMoteus: public JointBase
+    float moteus_kd = 0;
+    float soft_start_duration_ms = 1;
 };
+
```
Why white space change?

lucienpeach replied:
Followed formatting used in `moteus_protocol.h` and `joint_base.h`, but notice now that other files (such as `robot_base.h`) do not have whitespace here. Google Style guide on [namespace formatting](https://google.github.io/styleguide/cppguide.html#Namespace_Formatting) seems to suggest a line of vertical whitespace in this context, although the style guide also recommends limiting vertical whitespace when possible. 

Fine with either decision.

ShaneRozenLevy replied:
Either is fine. Generally isolated white space changes are frowned upon.

ShaneRozenLevy commented on include/kodlab_mjbots_sdk/mjbots_control_loop.h:
```diff
@@ -190,14 +191,23 @@ MjbotsControlLoop<log_type, input_type, robot_type>::MjbotsControlLoop(std::shar
+  // Open loop check
+  for(const auto & j : joints_moteus){
+      if(!options.open_loop && j->is_open_loop()){
+        LOG_FATAL("Open loop option not set, not receiving position and/or velocity");
```
Lets make this a little more clear. Maybe:

`ControlLoopOptions has open_loop=false, but joint moteus QueryCommand struct does not query position and or velocity data. `

ShaneRozenLevy commented on include/kodlab_mjbots_sdk/mjbots_control_loop.h:
```diff
@@ -190,14 +191,23 @@ MjbotsControlLoop<log_type, input_type, robot_type>::MjbotsControlLoop(std::shar
       options.print_torques,
-      options.send_pd_commands);
+      options.use_pd_commands,
+      options.open_loop
```
Why does the `MjbotsHardwareInterface` take in the open_loop bool since it is not used in the class?

ShaneRozenLevy commented on src/joint_moteus.cpp:
```diff
@@ -0,0 +1,294 @@
+
+// kDefaultQuery defintion
+const ::mjbots::moteus::QueryCommand JointMoteus::kDefaultQuery{
+  ::mjbots::moteus::Resolution::kInt8,
```
I feel like we should comment this so its readable to the user.

ShaneRozenLevy replied:
Once these are commented I will want to give this a look over with Diego to make sure we are happy with the standards.

ShaneRozenLevy commented on src/joint_moteus.cpp:
```diff
@@ -0,0 +1,294 @@
+  ::mjbots::moteus::Resolution::kInt32,
+  ::mjbots::moteus::Resolution::kInt16,
+  ::mjbots::moteus::Resolution::kInt16,
+  ::mjbots::moteus::Resolution::kIgnore,
```
what is kIgnore in comprehensive? That doesn't make sense.

lucienpeach replied:
The value associated with this kIgnore is "rezero_state", for which there is no documentation within the mjbots `reference.md` file. As such, I opted to set the resolution to kIgnore, as I was unsure what information this contains / how it would interact with the closed-loop control (if at all). 

Think that it would be better to include all fields, like suggested, but either want to know what this field does or perhaps change the name of the query.

ShaneRozenLevy replied:
I think thats fine, just add a comment.

ShaneRozenLevy commented on src/joint_moteus.cpp:
```diff
@@ -0,0 +1,294 @@
+}
+
+bool JointMoteus::is_open_loop() {
+  bool rv;
```
I know I initially gave you the way to use this function, but right now its super silly. Consider instead:

```
return (query_type_.position == ::mjbots::moteus::Resolution::kIgnore ||
  query_type_.velocity == ::mjbots::moteus::Resolution::kIgnore);
```

ShaneRozenLevy commented on src/joint_moteus.cpp:
```diff
@@ -0,0 +1,294 @@
+
+const ::mjbots::moteus::Mode JointMoteus::get_mode() const {
+  if(query_type_.mode == ::mjbots::moteus::Resolution::kIgnore){
+    LOG_WARN("Mode resolution is set to kIgnore, while attempting to log register");
```
Comment should not mention log. Instead maybe say "get value".

ShaneRozenLevy replied:
Should also mention this is in the `QueryCommand`,  resolution is also set in the data sent to the moteus.

ShaneRozenLevy reviewed (changes requested):
Much closer than last time. Comments are mostly just clean up and making everything super duper clear.

lucienpeach committed:
Comment resolution and related improvements

griffinaddison commented on src/joint_moteus.cpp:
```diff
@@ -0,0 +1,305 @@
+  }
+}
+
+float JointMoteus::get_servo_torque() const {
+  if(query_type_.torque == ::mjbots::moteus::Resolution::kIgnore){
+    LOG_WARN("Torque resolution in QueryCommand is set to kIgnore, while attempting to get value");
+    return 0;
+  } else {
+    return servo_torque_;
+  }
+}
```
this whole function shouldn't be here because query torque only handles measured torque. parent class joint_base already has a get servo torque function.
```suggestion
```

griffinaddison commented on include/kodlab_mjbots_sdk/joint_moteus.h:
```diff
@@ -103,129 +90,219 @@ class JointMoteus: public JointBase
+         * @return float
+         */
+        float get_velocity() const override;
+
+        /**
+         * @brief Get the torque command. Overrides function in joint_base.h
+         * 
+         * @return float
+         */
+        float get_servo_torque() const override;
```
See comment in cpp. get servo torque already defined in parent class, was throwing resolution warning when not part of query.

```suggestion
```

griffinaddison commented on src/joint_moteus.cpp:
```diff
@@ -0,0 +1,305 @@
+    return servo_torque_;
+  }
+}
+
```
```suggestion
```

jdcaporale committed:
Apply suggestions from code review

Remove accidentally added function

Co-authored-by: Griffin Addison <82300925+griffinaddison@users.noreply.github.com>

jdcaporale reviewed (approved):
Small change added, but otherwise this has been a long time coming. Great job!
